File: Python/beginner/01_Sorting/MergeSort.py
import random

def merging(arr: list[int], left: int, mid: int, right: int) -> None:
    tmp = [0] * len(arr)

    i = left
    j = mid + 1
    k = left

    while i <= mid and j <= right:
        if arr[i] <= arr[j]:
            tmp[k] = arr[i]
            i += 1
        else:
            tmp[k] = arr[j]
            j += 1
        k += 1

    if i > mid:
        for a in range(j, right + 1):
            tmp[k] = arr[a]
            k += 1
    else:
        for a in range(i, mid + 1):
            tmp[k] = arr[a]
            k += 1
    
    for a in range(left, right + 1):
        arr[a] = tmp[a]


def merge_sort(arr: list[int], left: int, right: int) -> None:
    if left < right:
        mid = (left + right) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merging(arr, left, mid, right)

if __name__ == '__main__':
    arr = [random.randint(1, 100) for _ in range(10)]

    print("[ 생성된 배열 ]")
    print(*arr)

    merge_sort(arr, 0, len(arr) - 1)

    print("[ 정렬된 배열 ]")
    print(*arr)

File: Cpp/boj/Beginner/01_Sorting/MergeSort.cpp
//
// Created by seungsu on 22. 11. 22.
//

#include <iostream>
#include <vector>
#include <random>

using namespace std;

constexpr int MAX = 10;

static void print_arr(vector<int>& arr) {
    for (int k : arr)
        cout << k << " ";
    cout << "\n";
}

static void merging(vector<int>& arr, int left, int mid, int right) {
    vector<int> tmp(MAX);

    int i = left;
    int j = mid + 1;
    int k = left;

    while (i <= mid && j <= right) {
        if (arr[i] < arr[j]) tmp[k++] = arr[i++];
        else tmp[k++] = arr[j++];
    }

    if (i <= mid) {
        for (int a = i; a <= mid; a++)
            tmp[k++] = arr[a];
    }

    else {
        for (int a = j; a <= right; a++)
            tmp[k++] = arr[a];
    }

    for (int a = left; a <= right; a++)
        arr[a] = tmp[a];
}

static void merge_sort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        merge_sort(arr, left, mid);
        merge_sort(arr, mid + 1, right);
        merging(arr, left, mid, right);
    }
}

void merge_sort_test() {
    random_device rd;
    mt19937 rand(rd());
    uniform_int_distribution<int> dist(1, 100);

    vector<int> v(MAX);
    for (int i = 0; i < MAX; i++)
        v[i] = dist(rand);

    cout << "[ 생성된 배열 ]" << "\n";
    print_arr(v);

    merge_sort(v, 0, v.size() - 1);

    cout << "[ 정렬된 배열 ]" << "\n";
    print_arr(v);
}